presto-main-base/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -307,48 +311,141 @@ private RelationPlan addColumnMasks(Table table, RelationPlan plan, SqlPlannerCo
+                    tableArgument.isPruneWhenEmpty(),
+                    new PassThroughSpecification(tableArgument.isPassThroughColumns(), passThroughColumns.build()),
+                    requiredColumns,
+                    specification));
+        }
+    }
+
+    private static void addPassthroughColumns(ImmutableList.Builder<VariableReferenceExpression> outputVariables,
+                                         Analysis.TableArgumentAnalysis tableArgument, RelationPlan sourcePlan,
+                                         Optional<DataOrganizationSpecification> specification,
+                                         ImmutableList.Builder<PassThroughColumn> passThroughColumns,
+                                         PlanBuilder sourcePlanBuilder)
+    {
+        if (tableArgument.isPassThroughColumns()) {
+            // the original output symbols from the source node, not coerced
+            // note: hidden columns are included. They are present in sourcePlan.fieldMappings
+            outputVariables.addAll(sourcePlan.getFieldMappings());
+            Set<VariableReferenceExpression> partitionBy = specification
+                    .map(DataOrganizationSpecification::getPartitionBy)
+                    .map(ImmutableSet::copyOf)
+                    .orElse(ImmutableSet.of());
+            sourcePlan.getFieldMappings().stream()
+                    .map(variable -> new PassThroughColumn(variable, partitionBy.contains(variable)))
+                    .forEach(passThroughColumns::add);
+        }
+        else if (tableArgument.getPartitionBy().isPresent()) {
+            tableArgument.getPartitionBy().get().stream()
+                    // the original symbols for partitioning columns, not coerced
+                    .map(sourcePlanBuilder::translate)
+                    .forEach(variable -> {
+                        outputVariables.add(variable);
+                        passThroughColumns.add(new PassThroughColumn(variable, true));
+                    });
+        }
+    }
+
+    private static void mapColumns(Analysis.TableArgumentAnalysis tableArgument,
+                                   RelationPlan sourcePlan,
+                                   ImmutableMultimap.Builder<String, VariableReferenceExpression> columnMapping,
+                                   PlanBuilder sourcePlanBuilder)
+    {
+        RelationType sourceDescriptor = sourcePlan.getDescriptor();
+        for (int i = 0; i < sourceDescriptor.getAllFieldCount(); i++) {
+            Optional<String> name = sourceDescriptor.getFieldByIndex(i).getName();
+            if (name.isPresent()) {
+                columnMapping.put(name.get(), sourcePlan.getVariable(i));
+                Optional<List<Expression>> partitionBy = tableArgument.getPartitionBy();
+                if (partitionBy.isPresent()) {
+                    sourcePlanBuilder.getTranslations().put(partitionBy.get().get(i), sourcePlan.getVariable(i));
```

Comment:
Is this correct? the parition by is getting an "i" which is not even related to it, its related to field count